src/thermo/ConstDensityTabulatedThermo.cpp

```diff
@@ -0,0 +1,203 @@
+/**
+ *  @file ConstDensityTabulatedThermo.cpp
+ * Definition file for a derived class of ThermoPhase that assumes either
+ * an ideal gas or ideal solution approximation and handles
+ * variable pressure standard state methods for calculating
+ * thermodynamic properties (see \ref thermoprops and
+ * class \link Cantera::ConstDensityTabulatedThermo ConstDensityTabulatedThermo\endlink).
+ */
+/*
+ * Copyright (2005) Sandia Corporation. Under the terms of
+ * Contract DE-AC04-94AL85000 with Sandia Corporation, the
+ * U.S. Government retains certain rights in this software.
+ */
+
+#include "cantera/thermo/ConstDensityTabulatedThermo.h"
+#include "cantera/base/ctml.h"
+
+namespace Cantera
+{
+
+void ConstDensityTabulatedThermo::setMoleFractions(const doublereal* const x)
+{
+	Phase::setMoleFractions(x);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::setMoleFractions_NoNorm(const doublereal* const x)
+{
+	Phase::setMoleFractions_NoNorm(x);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::setMassFractions(const doublereal* const y)
+{
+	Phase::setMassFractions(y);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::setMassFractions_NoNorm(const doublereal* const y)
+{
+	Phase::setMassFractions_NoNorm(y);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::setConcentrations(const doublereal* const conc)
+{
+	Phase::setConcentrations(conc);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::_updateThermo()
+{
+	doublereal tnow = temperature();
+	doublereal xnow = moleFraction(m_kk_mod);
+	doublereal c[4];
+	doublereal dS_corr = 0.0;
+	doublereal tlow = 0.0, thigh = 0.0;
+	int type = 0;
+	if (m_tlast != tnow || m_xlast != xnow) {
+		c[0]=tnow;
+		c[1]=interp_h(xnow) * 1e3; // 1e3 for conversion J/mol -> J/kmol
+		if (xnow==0) {
+			dS_corr = -BigNumber;
+		} else if (xnow==1) {
+			dS_corr = BigNumber;
+		} else {
+			dS_corr = GasConstant*std::log(xnow/(1.0-xnow));
+		}
+		c[2]=interp_s(xnow) * 1e3 + dS_corr; // 1e3 for conversion J/K/mol -> J/K/kmol
+		c[3]=0.0;
+		type = m_spthermo.reportType(m_kk_mod);
+		tlow = m_spthermo.minTemp(m_kk_mod);
+		thigh = m_spthermo.maxTemp(m_kk_mod);
+		shared_ptr<SpeciesThermoInterpType> stit(
+				newSpeciesThermoInterpType(type, tlow, thigh, OneAtm, c));
+		m_spthermo.modifySpecies(m_kk_mod, stit);
+		m_spthermo.update(tnow, &m_cp0_R[0], &m_h0_RT[0],
+				&m_s0_R[0]);
+		for (size_t k = 0; k < m_kk; k++) {
+			m_g0_RT[k] = m_h0_RT[k] - m_s0_R[k];
+		}
+		m_xlast = xnow;
+		m_tlast = tnow;
+	}
+}
+
+void ConstDensityTabulatedThermo::initThermoXML(XML_Node& phaseNode, const std::string& id_)
+{
+	ConstDensityTabulatedThermo::initThermo();
+	size_t offset;
+	doublereal a=0,b=0,c=0;
+	std::ifstream t;
+	std::string line;
+	int count = 0;
+
+	if (phaseNode.hasChild("thermo")) {
+		XML_Node& thermoNode = phaseNode.child("thermo");
+		std::string model = thermoNode["model"];
+		if (model == "ConstDensityTabulatedThermo") {
+			if (thermoNode.hasChild("data")) {

```

Comment:
Please use spaces and not tabs to indent the code. Four spaces = one tab

Reply:
This file no longer exists in the PR.